File: src/controllers/game_controller.rs
pub mod game {
    use std::any;

    use actix_web::{web, post, get, HttpResponse, Responder};

    // use diesel::query_dsl::methods::{FilterDsl};
    use diesel::{allow_tables_to_appear_in_same_query, joinable, ExpressionMethods, JoinOnDsl, NullableExpressionMethods, QueryDsl, RunQueryDsl, SelectableHelper};
    use diesel_json::Json;
    use serde_json::json;

    use crate::db::lib::establish_connection;
    use crate::db::models::challenge_model::Challenge;
    use crate::db::models::game_model::{Game, GameModel};
    use crate::db::models::game_challenge_model::{GameChallenge, GameSimpleChallenge, InsertGameChallenge};
    use crate::db::schema::game_challenges as game_challenges_schema;
    use crate::db::schema::games as games_schema;
    use crate::db::schema::challenges as challenges_schema;

    use crate::UserClaims;

    use serde_derive::{Serialize, Deserialize};

    #[derive(Debug, Serialize, Deserialize)]
    struct ListGamePayload {
        size: i32,
        offset: Option<i32>
    }

    #[post("/api/admin/game/list")]
    pub async fn list(jwt_user: UserClaims, payload: web::Json<ListGamePayload>,) -> impl Responder {
        let connection = &mut establish_connection();
        let game_results = games_schema::dsl::games
            .select(Game::as_select())
            .offset(payload.offset.unwrap_or(0) as i64)
            .limit(payload.size as i64)
            .load::<Game>(connection);

        match game_results {
            Ok(data) => {
                return HttpResponse::Ok().json(json!({
                    "code": 200,
                    "data": json!(data.iter().map(|gm| {
                        json!({
                            "game_id": gm.game_id,
                            "name": gm.name,
                            "summary": gm.summary,
                            "start_time": gm.start_time,
                            "end_time": gm.end_time,
                            "visible": gm.visible,
                            "poster": gm.poster
                        })
                    }).collect::<Vec<_>>())
                }));
            },
            Err(_) => {
                return HttpResponse::InternalServerError().json(json!({
                    "code": 500,
                    "message": "Server error"
                }));
            }
        }
    }

    #[post("/api/admin/game/create")]
    pub async fn create(jwt_user: UserClaims, payload: web::Json<GameModel>,) -> impl Responder {

        let connection = &mut establish_connection();
        
        let game_model: Game = payload.into_inner().into();
        let values = vec![ game_model ];

        match diesel::insert_into(games_schema::dsl::games)
            .values(&values)
            .returning(Game::as_select())
            .get_result::<Game>(connection) {
                Ok(game) => HttpResponse::Ok().json(json!({
                    "code": 200,
                    "data": {
                        "game_id": game.game_id
                    }
                })),
                Err(e) => {
                    println!("{:?}", e);

                    HttpResponse::InternalServerError().json(json!({
                        "code": 500,
                        "message": "Failed to create challenge"
                    }))
                }
            }
    }

    #[get("/api/admin/game/{game_id}")]
    pub async fn get(jwt_user: UserClaims, payload_game_id: web::Path<String>,) -> impl Responder {
        let connection = &mut establish_connection();

        let mut game_id_parsed = 0;

        match payload_game_id.parse::<i64>() {
            Ok(num) => {
                game_id_parsed = num;
            },
            Err(_) => {
                return HttpResponse::BadRequest().json(json!({
                    "code": 400,
                    "message": "Game id error"
                }))
            }
        }
        
        let game_record_result = games_schema::dsl::games
            .filter(games_schema::dsl::game_id.eq(game_id_parsed))
            .first::<Game>(connection);

        // challenges_schema::dsl::challenges.on(on)

        let game_challenges_result: Result<Vec<(GameChallenge, Option<Challenge>)>, diesel::result::Error> = game_challenges_schema::dsl::game_challenges
            .inner_join(challenges_schema::dsl::challenges.on(game_challenges_schema::challenge_id.eq(challenges_schema::challenge_id)))
            .select((
                game_challenges_schema::all_columns,
                challenges_schema::all_columns.nullable()
            ))
            .filter(game_challenges_schema::game_id.eq(payload_game_id.parse::<i64>().unwrap()))
            .load::<(GameChallenge, Option<Challenge>)>(connection);

        match game_record_result {
            Ok(game_record) => {
                match game_challenges_result {
                    Ok(game_challenges) => {
                        let mut result: GameModel = game_record.into();
                        result.challenges = game_challenges.iter().map(|(gc, c)| {

                            let judge_config_data = if gc.judge_config.is_some() {
                                gc.judge_config.as_ref().unwrap().0.clone()
                            } else {
                                c.as_ref().unwrap().judge_config.as_ref().unwrap().0.clone()
                            };

                            return GameSimpleChallenge {
                                challenge_id: c.as_ref().unwrap().challenge_id,
                                challenge_name:c.as_ref().unwrap().name.clone(),
                                score: gc.score,
                                solved: gc.solved.len() as i32,
                                category: c.as_ref().unwrap().category.0.clone(),
                                judge_config: judge_config_data
                            }
                        }).collect::<Vec<_>>();

                        HttpResponse::Ok().json(json!({
                            "code": 200,
                            "data": result
                        }))
                    },
                    Err(_) => {
                        HttpResponse::InternalServerError().json(json!({
                            "code": 500,
                            "message": "Server error"
                        }))
                    }
                }
            },
            Err(e) => {
                println!("{:?}", e);
                HttpResponse::InternalServerError().json(json!({
                    "code": 500,
                    "message": "Server error"
                }))
            }
        }
    }

    #[derive(Debug, Serialize, Deserialize)]
    struct AddGameChallengePayload {
        game_id: i64,
        challenge_id: i64
    }

    #[post("/api/admin/game/challenge/add")]
    pub async fn add_challenge(jwt_user: UserClaims, payload: web::Json<AddGameChallengePayload>,) -> impl Responder {
        let connection = &mut establish_connection();

        let game_record_result = games_schema::dsl::games
            .filter(games_schema::dsl::game_id.eq(payload.game_id))
            .first::<Game>(connection);

        match game_record_result {
            Ok(game_record) => {
                let challenge_record_result = challenges_schema::dsl::challenges
                    .filter(challenges_schema::dsl::challenge_id.eq(payload.challenge_id))
                    .first::<Challenge>(connection);
                match challenge_record_result {
                    Ok(challenge_record) => {

                        let new_value = InsertGameChallenge {
                            game_id: game_record.game_id,
                            challenge_id: payload.challenge_id,
                            score: 500f64,
                            enabled: false,
                            solved: Json::new(vec![]),
                            hints: Some(vec![]),
                            judge_config: Some(Json::new(challenge_record.judge_config.as_ref().unwrap().0.clone())),
                            belong_stage: None
                        };

                        let value_list = vec![ new_value ];

                        match diesel::insert_into(game_challenges_schema::dsl::game_challenges)
                            .values(value_list)
                            .returning(GameChallenge::as_select())
                            .get_result::<GameChallenge>(connection) {
                            Ok(v) => {
                                HttpResponse::Ok().json(json!({
                                    "code": 200,
                                    "data": json![{
                                        "ingame_id": v.ingame_id
                                    }]
                                }))
                            },
                            Err(_) => {
                                HttpResponse::InternalServerError().json(json!({
                                    "code": 500,
                                    "message": "Server error"
                                }))
                            }
                        }


                    },
                    Err(_) => {
                        return HttpResponse::BadRequest().json(json!({
                            "code": 401,
                            "message": "Bad request"
                        }))
                    }
                }
            },
            Err(_) => {
                return HttpResponse::BadRequest().json(json!({
                    "code": 401,
                    "message": "Bad request"
                }))
            }
        }
    }

}

File: src/controllers/game_controller.go
package controllers

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"a1ctf/src/db/models"
	dbtool "a1ctf/src/utils/db_tool"
)

type ListGamePayload struct {
	Size   int `json:"size" binding:"min=0"`
	Offset int `json:"offset"`
}

type AddGameChallengePayload struct {
	GameID      int64 `json:"game_id" binding:"min=0"`
	ChallengeID int64 `json:"challenge_id" binding:"min=0"`
}

func ListGames(c *gin.Context) {
	var payload ListGamePayload
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": err.Error(),
		})
		return
	}

	var games []models.Game
	query := dbtool.DB().Offset(payload.Offset).Limit(payload.Size)

	if err := query.Find(&games).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "Failed to load games",
		})
		return
	}

	data := make([]gin.H, 0, len(games))
	for _, game := range games {
		data = append(data, gin.H{
			"game_id":    game.GameID,
			"name":       game.Name,
			"summary":    game.Summary,
			"start_time": game.StartTime,
			"end_time":   game.EndTime,
			"visible":    game.Visible,
			"poster":     game.Poster,
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 200,
		"data": data,
	})
}

func CreateGame(c *gin.Context) {
	var payload models.Game
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": err.Error(),
		})
		return
	}

	game := models.Game{
		Name:                 payload.Name,
		Summary:              payload.Summary,
		StartTime:            payload.StartTime,
		EndTime:              payload.EndTime,
		Visible:              payload.Visible,
		Poster:               payload.Poster,
		WpExpireTime:         payload.WpExpireTime,
		Stages:               payload.Stages,
		RequireWp:            payload.RequireWp,
		ContainerNumberLimit: payload.ContainerNumberLimit,
		TeamNumberLimit:      payload.TeamNumberLimit,
		PracticeMode:         payload.PracticeMode,
		InviteCode:           payload.InviteCode,
		Description:          payload.Description,
	}

	if err := dbtool.DB().Create(&game).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "Failed to create game",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 200,
		"data": gin.H{
			"game_id": game.GameID,
		},
	})
}

func GetGame(c *gin.Context) {
	gameIDStr := c.Param("game_id")
	gameID, err := strconv.ParseInt(gameIDStr, 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": "Invalid game ID",
		})
		return
	}

	var game models.Game
	if err := dbtool.DB().Where("game_id = ?", gameID).First(&game).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{
				"code":    404,
				"message": "Game not found",
			})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{
				"code":    500,
				"message": "Failed to load game",
			})
		}
		return
	}

	var gameChallenges []struct {
		models.GameChallenge
		models.Challenge
	}

	// 使用 Preload 进行关联查询
	if err := dbtool.DB().Table("game_challenges").
		Joins("LEFT JOIN challenges ON game_challenges.challenge_id = challenges.challenge_id").
		Where("game_id = ?", gameID).
		Scan(&gameChallenges).Error; err != nil {

		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "Failed to load game challenges",
		})
		return
	}

	result := gin.H{
		"game_id":                game.GameID,
		"name":                   game.Name,
		"summary":                game.Summary,
		"description":            game.Description,
		"poster":                 game.Poster,
		"invite_code":            game.InviteCode,
		"start_time":             game.StartTime,
		"end_time":               game.EndTime,
		"practice_mode":          game.PracticeMode,
		"team_number_limit":      game.TeamNumberLimit,
		"container_number_limit": game.ContainerNumberLimit,
		"require_wp":             game.RequireWp,
		"wp_expire_time":         game.WpExpireTime,
		"stages":                 game.Stages,
		"visible":                game.Visible,
		"challenges":             make([]gin.H, 0, len(gameChallenges)),
	}

	for _, gc := range gameChallenges {
		judgeConfig := gc.GameChallenge.JudgeConfig
		if judgeConfig == nil {
			judgeConfig = gc.Challenge.JudgeConfig
		}

		result["challenges"] = append(result["challenges"].([]gin.H), gin.H{
			"challenge_id":   gc.Challenge.ChallengeID,
			"challenge_name": gc.Challenge.Name,
			"total_score":    gc.GameChallenge.TotalScore,
			"cur_score":      gc.GameChallenge.CurScore,
			"hints":          gc.GameChallenge.Hints,
			"solve_count":    len(gc.Solved),
			"category":       gc.Challenge.Category,
			"judge_config":   judgeConfig,
			"belong_stage":   gc.GameChallenge.BelongStage,
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 200,
		"data": result,
	})
}

func AddGameChallenge(c *gin.Context) {
	var payload AddGameChallengePayload
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    400,
			"message": err.Error(),
		})
		return
	}

	var game models.Game
	if err := dbtool.DB().Where("game_id = ?", payload.GameID).First(&game).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{
				"code":    404,
				"message": "Game not found",
			})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{
				"code":    500,
				"message": "Failed to verify game",
			})
		}
		return
	}

	var challenge models.Challenge
	if err := dbtool.DB().Where("challenge_id = ?", payload.ChallengeID).First(&challenge).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{
				"code":    404,
				"message": "Challenge not found",
			})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{
				"code":    500,
				"message": "Failed to verify challenge",
			})
		}
		return
	}

	gameChallenge := models.GameChallenge{
		GameID:      payload.GameID,
		ChallengeID: payload.ChallengeID,
		TotalScore:  500,
		CurScore:    500,
		Enabled:     false,
		Solved:      models.Solves{},
		Hints:       &models.StringArray{},
		JudgeConfig: challenge.JudgeConfig,
	}

	if err := dbtool.DB().Create(&gameChallenge).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    500,
			"message": "Failed to add challenge to game",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code": 200,
		"data": gin.H{
			"ingame_id": gameChallenge.IngameID,
		},
	})
}
